### PR TITLE
[FLINK-7008] [cep] Update NFA state only when the NFA changes

### PR DESCRIPTION
**Pull request by dianfu**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by dianfu** [comment]
Hi @kl0u @dawidwys, could you help to review this PR when you are free? Thanks a lot.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java [hunk 1]
```diff
@@ -324,6 +327,83 @@ public boolean filter(Event value) throws Exception {
...
+				return value.getName().equals("e");
+			}
+		}).within(Time.milliseconds(10));
+
+		NFACompiler.NFAFactory<Event> nfaFactory = NFACompiler.compileFactory(pattern, Event.createTypeSerializer(), true);
+		NFA<Event> nfa = nfaFactory.createNFA();
+		nfa.process(new Event(1, "b", 1.0), 1L);
+		assertFalse(nfa.isNFAChanged());
+
+		nfa.nfaChanged = false;
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java]: Could you add comments for each assertion, what situation in NFA it covers?

Comment (dianfu) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java]: @dawidwys thanks a lot for the review. Have added more comments.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java [hunk 2]
```diff
@@ -324,6 +328,89 @@ public boolean filter(Event value) throws Exception {
 		}
 	}
 
+	@Test
+	public void testNFAChange() {
+		Pattern<Event, ?> pattern = Pattern.<Event>begin("start").where(new SimpleCondition<Event>() {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java]: I feel the Pattern is a bit to complicated for this test. e.g. the notPattern does not introduce any new corner cases.

I miss a test for situation where only one CS is generated. You can simulate it with:

    @Test
    public void testNFAChangedOnOneNewComputationState() {
    	Pattern<Event, ?> pattern = Pattern.<Event>begin("start").where(new SimpleCondition<Event>() {
    		@Override
    		public boolean filter(Event value) throws Exception {
    			return value.getName().equals("start");
    		}
    	}).followedBy("a*").where(new SimpleCondition<Event>() {
    		private static final long serialVersionUID = 1858562682635302605L;

    		@Override
    		public boolean filter(Event value) throws Exception {
    			return value.getName().equals("a");
    		}
    	}).oneOrMore().optional().next("end").where(new IterativeCondition<Event>() {
    		private static final long serialVersionUID = 8061969839441121955L;

    		@Override
    		public boolean filter(Event value, Context<Event> ctx) throws Exception {
    			return value.getName().equals("b");
    		}
    	}).within(Time.milliseconds(10));

    	NFACompiler.NFAFactory<Event> nfaFactory = NFACompiler.compileFactory(pattern, Event.createTypeSerializer(), true);
    	NFA<Event> nfa = nfaFactory.createNFA();

    	// both the queue of ComputationStatus and eventSharedBuffer have changed
    	nfa.nfaChanged = false;
    	nfa.process(new Event(6, "start", 1.0), 6L);
    	nfa.nfaChanged = false;
    	nfa.process(new Event(6, "a", 1.0), 7L);
    	assertTrue("NFA status should change as the event matches the take condition of 'middle2' state", nfa.isNFAChanged());
    }

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java [hunk 3]
```diff
@@ -324,6 +328,89 @@ public boolean filter(Event value) throws Exception {
...
+			private static final long serialVersionUID = 1858562682635302605L;
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java]: Anyway I feel it would be easier to read and understand if we extract a test for a `nfaChange` case at a time.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java [hunk 4]
```diff
@@ -270,10 +270,12 @@ private void updateLastSeenWatermark(long timestamp) {
...
 	private void updateNFA(NFA<IN> nfa) throws IOException {
-		if (nfa.isEmpty()) {
-			nfaOperatorState.clear();
-		} else {
-			nfaOperatorState.update(nfa);
+		if (nfa.isNFAChanged()) {
+			if (nfa.isEmpty()) {
+				nfaOperatorState.clear();
+			} else {
+				nfaOperatorState.update(nfa);
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java]: An IT test for it would be handy.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 5]
```diff
@@ -152,6 +153,12 @@
 
 	private TypeSerializer<T> eventSerializer;
 
+	/**
+	 * Flag indicating whether the matching status of the state machine has changed.
+	 */
+	@VisibleForTesting
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: I think we need setter anyway (see AbstractKeyedCEPPatternOperator:277). So I would make it private.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java [hunk 6]
```diff
@@ -237,8 +254,15 @@ public boolean isEmpty() {
 						computationState.getCounter());
 
 				newComputationStates = Collections.emptyList();
+				nfaChanged = true;
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Just a question/side comment. This case always happens in pair with pruning sharedBuffer right? (NFA:330)

I would still leave it here, cause it's better be safe than sorry.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java [hunk 7]
```diff
@@ -114,6 +115,28 @@ public DeweyNumber getVersion() {
 		return version;
 	}
 
+	@Override
+	public boolean equals(Object obj) {
+		if (obj instanceof ComputationState) {
+			ComputationState other = (ComputationState) obj;
+			return Objects.equals(state, other.state) &&
+				event == other.event &&
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java]: how about comparing events with `.equals()`? It is not a primitive object.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java (hunk 4 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java]: Currently for each input, we will deserialize the NFA and the `nfaChanged` field is `false` by default. So I think we don't need to do that. Thoughts?

Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java]: Hmm, right we do access it from `StateBackend` each time. Not sure though if each time it is deserialized. I don't think there is such assumption. @kl0u could you shed some more light on it?

If I am right and it is not deserialized each time, I would reset the `nfaChanged` before writing it into the `StateBackend`.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java (hunk 7 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java]: For `ComputationState`, I think two `ComputationState` equals only if the last taken events of them point to the same event. So `.equals()` may be not correct. Thoughts?

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java (hunk 4 above)
Comment (kl0u) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java]: This is only true for the `RocksDB` state backend. If you are using the `Heap` state backend, you do not (de)serialize at every access but you keep the same object. So the flag should be re-set.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 6 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: I think this may not always be true. For example, the `eventSharedBuffer.release`(NFA:250) may already have released the events that will be pruned. This will cause the pruning sharedBuffer do nothing. Thoughts?
Anyway, leave it here will be more safe.:)

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java (hunk 7 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java]: The combination of `Object.equals(event, other.event)` and `count == other.counter` tells exactly that. I would not base the implementation on reference equity.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java (hunk 4 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/operator/AbstractKeyedCEPPatternOperator.java]: Make sense.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java (hunk 7 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/ComputationState.java]: Make sense.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 6 above)
Comment (dawidwys) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: Yes, you are right! I think then a test for that case would also be nice ;)

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java (hunk 2 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java]: Nice catch, will add this test to the test case.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java (hunk 3 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFATest.java]: OK.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 5 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: OK.

File: flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java (hunk 6 above)
Comment (dianfu) [inline flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/NFA.java]: OK.

**Comment by dianfu** [comment]
@dawidwys @kl0u thanks a lot for the review comments, have updated the PR.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CEPOperatorTest.java [hunk 8]
```diff
@@ -297,6 +297,85 @@ public void testKeyedAdvancingTimeWithoutElements() throws Exception {
 	}
 
 	@Test
+	public void testKeyedCEPOperatorNFAChanged() throws Exception {
```
Comment (dawidwys) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CEPOperatorTest.java]: I meant rather checking if the resetting for the flag works correct. We could do it by veryfying correct number of invocations of `update` method. Also we should test it with both RocksDB and InMemory state backends.

We can do it with Mockito. Just a suggestion for the code:

    @Test
    public void testKeyedCEPOperatorNFAChanged() throws Exception {

    	String rocksDbPath = tempFolder.newFolder().getAbsolutePath();
    	RocksDBStateBackend rocksDBStateBackend = new RocksDBStateBackend(new MemoryStateBackend());
    	rocksDBStateBackend.setDbStoragePath(rocksDbPath);

    	KeyedCEPPatternOperator<Event, Integer> operator = new KeyedCEPPatternOperator<>(
    		Event.createTypeSerializer(),
    		true,
    		IntSerializer.INSTANCE,
    		new SimpleNFAFactory(),
    		true);
    	OneInputStreamOperatorTestHarness<Event, Map<String, List<Event>>> harness = getCepTestHarness(operator);

    	try {
    		harness.setStateBackend(rocksDBStateBackend);

    		harness.open();

    		final ValueState nfaOperatorState = Whitebox.<ValueState>getInternalState(operator, "nfaOperatorState");
    		final ValueState nfaOperatorStateSpy = Mockito.spy(nfaOperatorState);
    		Whitebox.setInternalState(operator, "nfaOperatorState", nfaOperatorStateSpy);

    		Event startEvent = new Event(42, "c", 1.0);
    		SubEvent middleEvent = new SubEvent(42, "a", 1.0, 10.0);
    		Event endEvent = new Event(42, "b", 1.0);

    		harness.processElement(new StreamRecord<>(startEvent, 1L));
    		harness.processElement(new StreamRecord<>(new Event(42, "d", 1.0), 4L));
    		harness.processElement(new StreamRecord<Event>(middleEvent, 4L));
    		harness.processElement(new StreamRecord<>(endEvent, 4L));

    		Mockito.verify(nfaOperatorStateSpy, Mockito.times(3)).update(Mockito.any());
    		// get and verify the output

    		Queue<Object> result = harness.getOutput();

    		assertEquals(1, result.size());

    		verifyPattern(result.poll(), startEvent, middleEvent, endEvent);
    	} finally {
    		harness.close();
    	}
    }

Comment (dianfu) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/operator/CEPOperatorTest.java]: Got it. Make sense and have updated the PR.

**Comment by dawidwys** [comment]
Thanks for the work @dianfu! It looks good now. Let's just wait for travis to give green on it.
